File: 0735-asteroid-collision/0735-asteroid-collision.cpp
class Solution {
public:
    vector<int> asteroidCollision(vector<int>& asteroids) {
        stack<int> st;

        for(int asteroid : asteroids) {
            int flag = 1;
            
            //check possible collisions
            while(!st.empty() && (st.top() > 0 && asteroid < 0)) {
                //top asteroid is smaller
                if (abs(st.top()) < abs(asteroid)) {
                    st.pop();
                    continue;
                }
                //asteroids are same size
                else if (abs(st.top()) == abs(asteroid)) {
                    st.pop();
                }
                //current asteroid is samller
                flag = 0;
                break;
            }

            if (flag) {
                st.push(asteroid);
            }
        }

        //add asteroids from stack to vector in reverse order
        vector<int> remaining (st.size());
        for(int i = remaining.size() - 1; i >= 0; i--) {
            remaining[i] = st.top();
            st.pop();
        }

        return remaining;
    } //runtime : O(N), space : O(N)
};

File: 0735-asteroid-collision/0735-asteroid-collision.py
class Solution:
    def asteroidCollision(self, asteroids: List[int]) -> List[int]:
        ans = []
        st = []
        for asteroid in asteroids:
            if asteroid>0:
                st.append(asteroid)
            else:
                #check negative big collisions
                while len(st)>0 and st[-1]<abs(asteroid):
                    st.pop()
                    
                #check total annilation 
                if len(st)==0:
                    ans.append(asteroid)
                    
                #negative small and equal collisions
                else:
                    if st[-1] == abs(asteroid):
                        st.pop()
        ans += st
        return ans
                    